In [1]:
import numpy as np
import pandas as pd

### 1. Загрузка и подготовка данных ###

Загрузим оба сета и посмотрим на признаки.

In [2]:
train_df = pd.read_csv('train.csv')
train_df.head()

item_id       user_id                 region              city  \
0  b912c3c6a6ad  e00f8ff2eaf9   Свердловская область      Екатеринбург   
1  2dac0150717d  39aeb48f0017      Самарская область            Самара   
2  ba83aefab5dc  91e2f88dd6e3     Ростовская область    Ростов-на-Дону   
3  02996f1dd2ea  bf5cccea572d              Татарстан  Набережные Челны   
4  7c90be56d2ab  ef50846afc0b  Волгоградская область         Волгоград   

  parent_category_name               category_name  \
0          Личные вещи  Товары для детей и игрушки   
1      Для дома и дачи           Мебель и интерьер   
2  Бытовая электроника               Аудио и видео   
3          Личные вещи  Товары для детей и игрушки   
4            Транспорт                  Автомобили   

                       param_1     param_2 param_3                  title  \
0    Постельные принадлежности         NaN     NaN  Кокоби(кокон для сна)   
1                       Другое         NaN     NaN      Стойка для Одежды   
2  Видео, DVD и Blu-ray плееры         NaN     NaN         Philips bluray   
3         Автомобильные кресла         NaN     NaN             Автокресло   
4                   С пробегом  ВАЗ (LADA)    2110         ВАЗ 2110, 2003   

                                         description    price  \
0  Кокон для сна малыша,пользовались меньше месяц...    400.0   
1          Стойка для одежды, под вешалки. С бутика.   3000.0   
2  В хорошем состоянии, домашний кинотеатр с blu ...   4000.0   
3                             Продам кресло от0-25кг   2200.0   
4                           Все вопросы по телефону.  40000.0   

   item_seq_number activation_date user_type  \
0                2      2017-03-28   Private   
1               19      2017-03-26   Private   
2                9      2017-03-20   Private   
3              286      2017-03-25   Company   
4                3      2017-03-16   Private   

                                               image  image_top_1  \
0  d10c7e016e03247a3bf2d13348fe959fe6f436c1caf64c...       1008.0   
1  79c9392cc51a9c81c6eb91eceb8e552171db39d7142700...        692.0   
2  b7f250ee3f39e1fedd77c141f273703f4a9be59db4b48a...       3032.0   
3  e6ef97e0725637ea84e3d203e82dadb43ed3cc0a1c8413...        796.0   
4  54a687a3a0fc1d68aed99bdaaf551c5c70b761b16fd0a2...       2264.0   

   deal_probability  
0           0.12789  
1           0.00000  
2           0.43177  
3           0.80323  
4           0.20797

In [3]:
test_df = pd.read_csv('test.csv')
test_df.head()

item_id       user_id                 region         city  \
0  6544e41a8817  dbe73ad6e4b5  Волгоградская область    Волгоград   
1  65b9484d670f  2e11806abe57   Свердловская область  Нижняя Тура   
2  8bab230b2ecd  0b850bbebb10  Новосибирская область       Бердск   
3  8e348601fefc  5f1d5c3ce0da    Саратовская область      Саратов   
4  8bd2fe400b89  23e2d97bfc7f   Оренбургская область      Бузулук   

  parent_category_name               category_name                 param_1  \
0          Личные вещи      Детская одежда и обувь           Для мальчиков   
1        Хобби и отдых                  Велосипеды                Дорожные   
2  Бытовая электроника               Аудио и видео  Телевизоры и проекторы   
3      Для дома и дачи             Бытовая техника               Для кухни   
4          Личные вещи  Товары для детей и игрушки         Детские коляски   

   param_2 param_3              title  \
0    Обувь      25    Отдам бесплатно   
1      NaN     NaN   Продам велосипед   
2      NaN     NaN                BBK   
3  Вытяжки     NaN  Вытяжка Jetair 60   
4      NaN     NaN  Коляска зима-лето   

                                         description    price  \
0                                       На ангарском      NaN   
1  Продам велосипед KAMA  F200,в нормальном состо...   3000.0   
2  Продам новый телевизор BBK  32 диагональ смарт...  15000.0   
3  Продам новую вытяжку в упаковке,с документами....   4500.0   
4  Продам отличную коляску. б/у 1 год. все вопрос...   4900.0   

   item_seq_number activation_date user_type  \
0               66      2017-04-18   Private   
1                4      2017-04-16   Private   
2               15      2017-04-17   Private   
3               70      2017-04-17   Private   
4               15      2017-04-15   Private   

                                               image  image_top_1  
0  a8b57acb5ab304f9c331ac7a074219aed4d349d8aef386...       2020.0  
1                                                NaN          NaN  
2  8c361112cb049745ef2d1b0ae73594fc5c107286b0c942...       2960.0  
3                                                NaN          NaN  
4  bc3cf6deef10840fc302e38eb48fa7748aa1e28d534f8f...       1002.0

In [4]:
print('Количество объектов в тренировочной выборке: ', train_df.shape[0])
print('Количество объектов в тестовой выборке: ', test_df.shape[0])

Количество объектов в тренировочной выборке:  1503424
Количество объектов в тестовой выборке:  508438


Попробуем поработать с одним признаком - самым информативным - текстовым описанием товара.

In [5]:
X_train = train_df['description']
y_train = train_df['deal_probability']
X_test_final = test_df['description']

Проверим, есть ли наны, и если да, заполним их пробелами.

In [6]:
True in X_train.isnull()

True

In [7]:
X_train = X_train.fillna(' ')
X_test_final = X_test_final.fillna(' ')

Разделим исходную тренировочную выборку на трэйн и тест в пропорции 70% - 30%.

In [8]:
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(X_train, y_train, test_size=0.3, random_state=42)

### 2. Baseline: LinearRegression.

Сделаем первый базовый сабмит.

Преобразуем текст в вектора с помощью TfidfVectorizer (практика показывает, что обычно он работает лучше, чем CountVectorizer). Будем использовать сырой текст. Нормализацию в этом проекте я делать, к сожалению, не буду: она занимает очень много времени и я уже не успею.
Стоп-слова возьмем из NLTK. 

In [9]:
from sklearn.feature_extraction.text import TfidfVectorizer
from nltk.corpus import stopwords

def vectorize_data(X_train, X_test, X_test_final):
    stops = stopwords.words('russian')
        
    tfidf = TfidfVectorizer(min_df=3, max_features=None, strip_accents='unicode', analyzer='word',
                            token_pattern=r'\w{1,}', ngram_range=(1, 2), use_idf=1, sublinear_tf=1, stop_words=stops)

    X_train_vec = tfidf.fit_transform(X_train)
    X_test_vec = tfidf.transform(X_test)
    X_test_final_vec = tfidf.transform(X_test_final)
    
    return X_train_vec, X_test_vec, X_test_final_vec

In [11]:
X_train1, X_test1, X_test_final1 = vectorize_data(X_train, X_test, X_test_final)

In [12]:
print('Количество объектов в тренировочной выборке: ', X_train1.shape[0])
print('Количество объектов в тестовой выборке: ', X_test1.shape[0])
print('Количество фичей: ', X_train1.shape[1])

Количество объектов в тренировочной выборке:  1052396
Количество объектов в тестовой выборке:  451028
Количество фичей:  1099263


Так как количество объектов и фичей огромное, а времени до дедлайна немного, то для того, чтобы успеть сделать минимум 6 самбитов, используем TruncatedSVD для снижения размерности до 200.

In [10]:
from sklearn.decomposition import TruncatedSVD

def truncate_features(X_train, X_test, X_test_final):
    tsvd = TruncatedSVD(n_components=200, random_state=42)

    X_train_tsvd = tsvd.fit_transform(X_train)
    X_test_tsvd = tsvd.transform(X_test)
    X_test_final_tsvd = tsvd.transform(X_test_final)
    
    return X_train_tsvd, X_test_tsvd, X_test_final_tsvd

In [16]:
X_train1_tsvd, X_test1_tsvd, X_test_final1_tsvd = truncate_features(X_train1, X_test1, X_test_final1)

Baseline-алгоритм - LinearRegression с дефолтными параметрами. Метрика - RMSE.

In [15]:
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error as mse
from math import sqrt

def train_alg(alg, X_train, y_train, X_test, y_test):
    alg.fit(X_train, y_train)

    y_pred = alg.predict(X_test)
    score = sqrt(mse(y_pred, y_test))

    print('RMSE: ', score)
    
    return alg

In [18]:
lr = LinearRegression(n_jobs=-1)
lr = train_alg(lr, X_train1_tsvd, y_train, X_test1_tsvd, y_test)

RMSE:  0.24685550665910516


In [23]:
y_pred_final1 = lr.predict(X_test_final1_tsvd)

In [18]:
def save_submission(y_pred_final, ids, file_name):
    for i in range(len(y_pred_final)):
        if y_pred_final[i] < 0:
            y_pred_final[i] = 0
            
    submit = pd.DataFrame(data={'item_id': ids, 'deal_probability': y_pred_final})
    submit.to_csv(file_name, index=False)

In [38]:
save_submission(y_pred_final1, test_df['item_id'], 'submit1.csv')

Первый сабмишн показал результат 0.2527.

### 3. Более сложные и быстрые алгоритмы.

В этой секции мы:
- приведем данные к нижнему регистру.
- обучим два быстрых регрессора с дефолтными параметрами: XGBRegressor из библиотеки py-xgboost и LGBMRegressor из библиотеки lightgbm.
- посмотрим, удастся ли улучшить финальный результат, и выберем лучший из регрессоров.

In [11]:
def normalize(text):
    text = text.lower()    
    
    return text

In [12]:
X_train_norm = X_train.apply(normalize)
X_test_norm = X_test.apply(normalize)
X_test_final_norm = X_test_final.apply(normalize)

In [13]:
X_train2, X_test2, X_test_final2 = vectorize_data(X_train_norm, X_test_norm, X_test_final_norm)
X_train2_tsvd, X_test2_tsvd, X_test_final2_tsvd = truncate_features(X_train2, X_test2, X_test_final2)

#### 3.1 XGBRegressor

In [16]:
from xgboost import XGBRegressor

xgb = XGBRegressor(seed=42)
xgb = train_alg(xgb, X_train2_tsvd, y_train, X_test2_tsvd, y_test)

RMSE:  0.24574451457112065


In [19]:
y_pred_final2 = xgb.predict(X_test_final2_tsvd)
save_submission(y_pred_final2, test_df['item_id'], 'submit2.csv')

Скор и на отложенной выборке, и на тестовой (0.2517) немного улучшился. Второй сабмишн поднял меня на 2 места.

#### 3.2 LGBMRegressor

In [20]:
import lightgbm as lgb

gbm = lgb.LGBMRegressor(objective='regression')
gbm.fit(X_train2_tsvd, y_train, eval_set=[(X_test2_tsvd, y_test)], eval_metric='l1', early_stopping_rounds=5)
y_pred = gbm.predict(X_test2_tsvd, num_iteration=gbm.best_iteration_)

print('RMSE: ', sqrt(mse(y_test, y_pred)))

[1]	valid_0's l1: 0.188465
Training until validation scores don't improve for 5 rounds.
[2]	valid_0's l1: 0.186978
[3]	valid_0's l1: 0.185697
[4]	valid_0's l1: 0.184661
[5]	valid_0's l1: 0.183719
[6]	valid_0's l1: 0.182861
[7]	valid_0's l1: 0.182086
[8]	valid_0's l1: 0.181406
[9]	valid_0's l1: 0.180756
[10]	valid_0's l1: 0.180169
[11]	valid_0's l1: 0.179634
[12]	valid_0's l1: 0.179114
[13]	valid_0's l1: 0.178676
[14]	valid_0's l1: 0.17823
[15]	valid_0's l1: 0.177816
[16]	valid_0's l1: 0.177432
[17]	valid_0's l1: 0.177087
[18]	valid_0's l1: 0.176764
[19]	valid_0's l1: 0.17647
[20]	valid_0's l1: 0.176215
[21]	valid_0's l1: 0.17597
[22]	valid_0's l1: 0.17574
[23]	valid_0's l1: 0.175519
[24]	valid_0's l1: 0.175262
[25]	valid_0's l1: 0.175018
[26]	valid_0's l1: 0.174803
[27]	valid_0's l1: 0.174598
[28]	valid_0's l1: 0.174432
[29]	valid_0's l1: 0.174255
[30]	valid_0's l1: 0.17408
[31]	valid_0's l1: 0.173915
[32]	valid_0's l1: 0.173761
[33]	valid_0's l1: 0.173575
[34]	valid_0's l1: 0.173427
[

In [21]:
y_pred_final3 = gbm.predict(X_test_final2_tsvd)
save_submission(y_pred_final3, test_df['item_id'], 'submit3.csv')

Скор и на отложенной выборке, и на тестовой (0.2484) увеличился на большее количество пунктов. Третий самбишн поднял меня на 11 мест.

### 4. Новые признаки
Попробуем добавить к исходным фичам три новые: регион, название раздела и название категории.
Это категориальные признаки, но мы сделаем их бинарными с помощью функции get_dummies, а потом объединим признаки с помощью numpy.hstack.

После этого натренируем на новых данных лучший регрессор из уже опробованных = LGBMRegressor (все еще с дефолтными параметрами).

In [51]:
other_features = train_df[['region', 'parent_category_name', 'category_name']]
other_features = pd.get_dummies(other_features)

In [52]:
other_features.shape

(1503424, 84)

In [53]:
X_train_oth, X_test_oth, y_train_oth, y_test_oth = train_test_split(other_features, train_df['deal_probability'],
                                                                    test_size=0.3, random_state=42)

In [54]:
X_train_new = np.hstack((X_train_oth, X_train2_tsvd))
X_test_new = np.hstack((X_test_oth, X_test2_tsvd))

In [55]:
gbm.fit(X_train_new, y_train, eval_set=[(X_test_new, y_test)], eval_metric='l1', early_stopping_rounds=5)
y_pred = gbm.predict(X_test_new, num_iteration=gbm.best_iteration_)

print('RMSE: ', sqrt(mse(y_test, y_pred)))

[1]	valid_0's l1: 0.186951
Training until validation scores don't improve for 5 rounds.
[2]	valid_0's l1: 0.184226
[3]	valid_0's l1: 0.182043
[4]	valid_0's l1: 0.18019
[5]	valid_0's l1: 0.178587
[6]	valid_0's l1: 0.177204
[7]	valid_0's l1: 0.175997
[8]	valid_0's l1: 0.17482
[9]	valid_0's l1: 0.173872
[10]	valid_0's l1: 0.172969
[11]	valid_0's l1: 0.172098
[12]	valid_0's l1: 0.171386
[13]	valid_0's l1: 0.170708
[14]	valid_0's l1: 0.170094
[15]	valid_0's l1: 0.169535
[16]	valid_0's l1: 0.169018
[17]	valid_0's l1: 0.168546
[18]	valid_0's l1: 0.168128
[19]	valid_0's l1: 0.167767
[20]	valid_0's l1: 0.167445
[21]	valid_0's l1: 0.167135
[22]	valid_0's l1: 0.166883
[23]	valid_0's l1: 0.16666
[24]	valid_0's l1: 0.166348
[25]	valid_0's l1: 0.166142
[26]	valid_0's l1: 0.165953
[27]	valid_0's l1: 0.165729
[28]	valid_0's l1: 0.165543
[29]	valid_0's l1: 0.165367
[30]	valid_0's l1: 0.165208
[31]	valid_0's l1: 0.165071
[32]	valid_0's l1: 0.164942
[33]	valid_0's l1: 0.164826
[34]	valid_0's l1: 0.16468


In [56]:
other_features_test = test_df[['region', 'parent_category_name', 'category_name']]
other_features_test = pd.get_dummies(other_features_test)

X_test_final_new = np.hstack((other_features_test, X_test_final2_tsvd))

In [57]:
y_pred_final4 = gbm.predict(X_test_final_new)
save_submission(y_pred_final4, test_df['item_id'], 'submit4.csv')

Скор на обеих выборках значительно улучшился. Результат на тестовой - 0.2411. Этот самбишн поднял меня на 62 места.
Получается, добавление новых признаков очень хорошо повлияло на результат.

### 5. Масштабирование данных

На результат некоторых моделей может очень сильно влиять то, отмасштабированы данные или нет.
Попробуем отмасштабировать наши данные с помощью StandardScaler и снова обучить LGBMRegressor.

In [64]:
from sklearn.preprocessing import StandardScaler

scaler = StandardScaler()

X_train_scal = scaler.fit_transform(X_train_new)
X_test_scal = scaler.fit_transform(X_test_new)
X_test_final_scal = scaler.fit_transform(X_test_final_new)

In [65]:
gbm.fit(X_train_scal, y_train, eval_set=[(X_test_scal, y_test)], eval_metric='l1', early_stopping_rounds=5)
y_pred = gbm.predict(X_test_scal, num_iteration=gbm.best_iteration_)

print('RMSE: ', sqrt(mse(y_test, y_pred)))

[1]	valid_0's l1: 0.186943
Training until validation scores don't improve for 5 rounds.
[2]	valid_0's l1: 0.184197
[3]	valid_0's l1: 0.182009
[4]	valid_0's l1: 0.180146
[5]	valid_0's l1: 0.178671
[6]	valid_0's l1: 0.177282
[7]	valid_0's l1: 0.176146
[8]	valid_0's l1: 0.174968
[9]	valid_0's l1: 0.174047
[10]	valid_0's l1: 0.173078
[11]	valid_0's l1: 0.17225
[12]	valid_0's l1: 0.171477
[13]	valid_0's l1: 0.170922
[14]	valid_0's l1: 0.170381
[15]	valid_0's l1: 0.169874
[16]	valid_0's l1: 0.169327
[17]	valid_0's l1: 0.168885
[18]	valid_0's l1: 0.168622
[19]	valid_0's l1: 0.168212
[20]	valid_0's l1: 0.167935
[21]	valid_0's l1: 0.167714
[22]	valid_0's l1: 0.167479
[23]	valid_0's l1: 0.167235
[24]	valid_0's l1: 0.167021
[25]	valid_0's l1: 0.166809
[26]	valid_0's l1: 0.166562
[27]	valid_0's l1: 0.166363
[28]	valid_0's l1: 0.166207
[29]	valid_0's l1: 0.166035
[30]	valid_0's l1: 0.165867
[31]	valid_0's l1: 0.165685
[32]	valid_0's l1: 0.165607
[33]	valid_0's l1: 0.165474
[34]	valid_0's l1: 0.1653

In [66]:
y_pred_final5 = gbm.predict(X_test_final_scal)
save_submission(y_pred_final5, test_df['item_id'], 'submit5.csv')

Скор на обеих выборках немного ухудшился (на тестовой - 0.2412). Получается, в нашем случае масштабирование ни на что не влияет и даже портит картину, поэтому лучше обойтись без него.

### 6. Поиск гиперпараметров

Подберем некоторые гиперпараметры для нашего регрессора. Для поиска будем использовать RandomizedSearchCV, т.к. он быстрее, чем GridSearchCV, и обычно не уступает в качестве.

Согласно рекомендациям здесь: https://lightgbm.readthedocs.io/en/latest/Parameters-Tuning.html (раздел For Better Accuracy), будем тюнить параметры learning_rate и num_leaves. Дефолтные значения - 0.1 для первого и 31 для второго. Также изменим некоторые другие параметры: поставим boosting='dart' (вместо 'gbdt') и увеличим число итераций до 200 (дефолт - 100).

In [82]:
from sklearn.model_selection import RandomizedSearchCV

def parameters_search(X_train, y_train):
    rand_results, rand_scores = [], []
    gbm = lgb.LGBMRegressor(objective='regression', num_iterations=200, boosting='dart')
    params = {'learning_rate': [0.001, 0.01, 0.1], 'num_leaves': range(31, 131, 10)}   
        
    rand = RandomizedSearchCV(gbm, param_distributions=params, scoring='neg_mean_squared_error', cv=5, 
                              n_iter=9, random_state=42)
    rand.fit(X_train, y_train)
    rand_results.append(rand.best_params_)
    rand_scores.append(rand.best_score_)
    
    print('Random Search results')
    print(pd.DataFrame(data={'Best score': rand_scores, 'Parameters': rand_results}))

In [81]:
import warnings
warnings.filterwarnings('ignore')

In [83]:
parameters_search(X_train_new, y_train)

Random Search results
   Best score                                 Parameters
0   -0.055253  {'learning_rate': 0.1, 'num_leaves': 111}


Обучим регрессор с лучшими параметрами. Заодно увеличим число итераций до 500.

In [86]:
tuned_gbm = lgb.LGBMRegressor(objective='regression', num_iterations=500,
                             boosting='dart', learning_rate=0.1, num_leaves=111, verbosity=0)

tuned_gbm.fit(X_train_new, y_train, eval_set=[(X_test_new, y_test)], eval_metric='l1', early_stopping_rounds=5)
y_pred = tuned_gbm.predict(X_test_new, num_iteration=gbm.best_iteration_)

print('RMSE: ', sqrt(mse(y_test, y_pred)))

[1]	valid_0's l1: 0.18666
Training until validation scores don't improve for 5 rounds.
[2]	valid_0's l1: 0.183701
[3]	valid_0's l1: 0.181312
[4]	valid_0's l1: 0.179253
[5]	valid_0's l1: 0.177494
[6]	valid_0's l1: 0.175946
[7]	valid_0's l1: 0.174602
[8]	valid_0's l1: 0.173434
[9]	valid_0's l1: 0.172367
[10]	valid_0's l1: 0.171396
[11]	valid_0's l1: 0.170533
[12]	valid_0's l1: 0.169721
[13]	valid_0's l1: 0.169012
[14]	valid_0's l1: 0.168354
[15]	valid_0's l1: 0.167735
[16]	valid_0's l1: 0.167215
[17]	valid_0's l1: 0.166695
[18]	valid_0's l1: 0.166226
[19]	valid_0's l1: 0.1658
[20]	valid_0's l1: 0.165421
[21]	valid_0's l1: 0.165063
[22]	valid_0's l1: 0.164749
[23]	valid_0's l1: 0.164486
[24]	valid_0's l1: 0.164213
[25]	valid_0's l1: 0.163974
[26]	valid_0's l1: 0.163777
[27]	valid_0's l1: 0.163584
[28]	valid_0's l1: 0.163371
[29]	valid_0's l1: 0.163224
[30]	valid_0's l1: 0.163076
[31]	valid_0's l1: 0.16295
[32]	valid_0's l1: 0.16282
[33]	valid_0's l1: 0.162697
[34]	valid_0's l1: 0.162575
[

In [87]:
y_pred_final6 = tuned_gbm.predict(X_test_final_new)
save_submission(y_pred_final6, test_df['item_id'], 'submit6.csv')

Скор увеличился на обоих выборках. Финальный результат на тестовой - 0.2400. Последний самбишн поднял меня еще на 12 мест.

### 7. Выводы

Финальные результаты:

| Submission | Holdout RMSE | Test RMSE |
|------|------||------|
|1. LinearRegression + 1 feature | 0.247 | 0.2527 |
|2. XGBRegressor + 1 feature | 0.246 | 0.2517 |
|3. LGBMRegressor + 1 feature | 0.242 | 0.2484 |
|4. LGBMRegressor + 4 features | 0.236 | 0.2411 |
|5. LGBMRegressor + scaled data (4 features) | 0.236 | 0.2412 |
|6. tuned LGBMRegressor + non-scaled data (4 features) | 0.235 | 0.2400 |


#### Что еще можно было бы сделать?
- нормализовать текстовые данные
- добавить еще больше признаков (например, title, city, param_1)
- увеличить размерность TruncatedSVD
- перебрать больший набор гиперпараметров для регрессора
- попробовать настекать алгоритмы

Сабмишены: https://www.kaggle.com/c/avito-demand-prediction/submissions?sortBy=date&group=all&page=1